### PR TITLE
DAST support for FreeStyle (Website, Workflow, API) & Pipeline (Website,Workflow).

### PR DESCRIPTION
**Pull request by pvivekdev**
Support for DAST Automated Scan from Fortify On Demand Plugin (Free Style Template).
https://almoctane-ams.saas.microfocus.com/ui/entity-navigation?p=112082/7001&entityType=work_item&id=888034
https://almoctane-ams.saas.microfocus.com/ui/entity-navigation?p=112082/7001&entityType=work_item&id=993003

Added Fod On Demand Dynamic Scan Post build plugin, 
Added FreeStyle Support DAST Scan types (Website, Workflow Driven, API).
Added pipeline Support for DAST (Website, Workflow)
Included Scope settings, Scan Type specific settings, Support to Create Application & Release for DAST. 
Included Network Auth/ Login Macro Setting to support Website/Workflow Driven. 
Included HTML File Type control to upload file into FOD API. 

### Testing done
1. Selection of Application and Release associated with current user.
2.  On selection of Fod On Demand Dynamic Post build action loads DAST UI controls. 
3. On selection of Scan type - Scan type specific UI controls are loaded. 
4. Saving the specific scan settings into DAST Fod API.
5. Loading the saved scan settings back into Jenkin UI controls on click on Jenkin Job - Save/Apply
6. Submitting Website and Workflow scan setting and initiate on click of Jenkin Job-  Build. 

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x ] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/fodupload/DynamicAssessmentBuildStep.java [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+import org.kohsuke.stapler.verb.POST;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.UUID;
+
+public class DynamicAssessmentBuildStep extends Recorder implements SimpleBuildStep {
```
Comment (pvivekdev) [inline src/main/java/org/jenkinsci/plugins/fodupload/DynamicAssessmentBuildStep.java:34]: Folder name -> DastFreeStyleBuildStep | DastPipelineBuildStep

File: src/main/webapp/DastFreeStyle.js [hunk 2]
```diff
@@ -0,0 +1,1199 @@
...
+            apiScanSettingRows.hide();
+        } else {
+            apiScanSettingRows.show();
+        }
+    }
+
+    workflowScanSettingVisibility(isVisible) {
+
+        let workflowScanSettingRows = jq('.dast-workflow-setting');
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:198]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 3]
```diff
@@ -0,0 +1,1199 @@
...
+            workflowScanSettingRows.hide();
+        } else {
+            workflowScanSettingRows.show();
+        }
+    }
+
+    scanSettingsVisibility(isVisible) {
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:206]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 4]
```diff
@@ -0,0 +1,1199 @@
...
+                        console.log(item);
+                        let ident = arr[index];
+                        jq('#listWorkflowDrivenAllowedHostUrl').append("<li>" + "<input type='checkbox'" +
+                            " id=' " + ident + " ' name='" + ident + "'>" + arr[index] + "</li>");
+                        if (jq('#workflowMacroHosts').val() === null || undefined) {
+                            jq('#workflowMacroHosts').val(arr[index]);
+                        } else {
+
+                            let host = jq('#workflowMacroHosts').val();
+                            if (host !== null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:467]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 5]
```diff
@@ -0,0 +1,1199 @@
...
+    resetNetworkSettings() {
+
+        jq('#networkUsernameRow').find('input').val(undefined);
+        jq('#networkPasswordRow').find('input').val(undefined);
+        jq('#ddlNetworkAuthType').prop('selectedIndex', 0);
+        jq('#webSiteNetworkAuthSettingEnabledRow').find('input:checkbox:first').trigger('click');
+    }
+
+    setScanType() {
+        if (this.scanSettings !== undefined && this.scanSettings !== null) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:609]: I'm perfectly fine with this way, it's more explicitly, but if you'd like to clean these if statements up by using truthiness it's perfectly safe to do...

if (this.scanSettings)

File: src/main/webapp/DastFreeStyle.js [hunk 6]
```diff
@@ -0,0 +1,1199 @@
...
+        if (this.scanSettings !== null && this.scanSettings.restrictToDirectoryAndSubdirectories !== null) {
+            {
+                jq('#restrictScan').prop('checked', this.scanSettings.restrictToDirectoryAndSubdirectories);
+            }
+        }
+    }
+
+
+    setScanPolicy() {
+        if (this.scanSettings !== undefined && this.scanSettings.policy !== null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:714]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 7]
```diff
@@ -0,0 +1,1199 @@
...
+            jq('#dast-standard-scan-scope').hide();
+    }
+
+    onWorkflowDrivenHostChecked(event) {
+
+        let allowedHost = jq('#workflowMacroHosts').val();
+        if (event.target.checked) {
+            let hostToAdd = event.target.name; //name point to the host returned from FOD Patch API
+
+            if (allowedHost !== undefined || null) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:898]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 8]
```diff
@@ -0,0 +1,1199 @@
...
+                if (allowedHost.length > 0) {
+                    allowedHost = allowedHost + "," + hostToAdd;
+                } else
+                    allowedHost = hostToAdd;
+                jq('#workflowMacroHosts').val(allowedHost);
+            }
+        } else {
+            if (allowedHost !== undefined || null) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:906]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 9]
```diff
@@ -0,0 +1,1199 @@
...
+        }
+    }
+
+    onApiTypeChanged() {
+
+        jq('.uploadMessage').text('');
+        this.onSourceChange('none');
+        let selectedApiTypeValue = jq('#apiTypeList').val();
+
+        if (selectedApiTypeValue === null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:928]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js [hunk 10]
```diff
@@ -0,0 +1,1199 @@
...
+            jq('#dast-graphQL-url').show();
+            jq('#graphQlRadioSource').val(jq('#' + event.target.id).val());
+        } else {
+            jq('.sourceOptions').prop('checked', false);
+        }
+    }
+
+    apiTypeUserControlVisibility(apiType, isVisible) {
+
+        if ((isVisible !== undefined || null)) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:972]: Also fix this if statement like I commented above

File: src/main/webapp/DastFreeStyle.js (hunk 9 above)
Comment (c0d3m0nky) [reply src/main/webapp/DastFreeStyle.js:928]: to find any remaining you can use this regex in Edit > Find > Find in Files

(undefined|null) +\|\| +(undefined|null)

File: src/main/webapp/DastFreeStyle.js [hunk 11]
```diff
@@ -0,0 +1,1199 @@
...
+            this.openApiScanVisibility(false);
+            this.grpcScanVisibility(false);
+            this.graphQlScanVisibility(false);
+            this.postmanScanVisibility(false);
+            switch (apiType) {
+                case "openApi":
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:978]: Use constant

File: src/main/webapp/DastFreeStyle.js [hunk 12]
```diff
@@ -0,0 +1,1199 @@
...
+        jq('#dast-openApi').closest('.tr').hide()
+        jq('#dast-graphQL').closest('.tr').hide();
+        jq('#dast-grpc').closest('.tr').hide();
+    }
+
+    apiScanSettingVisibility(isVisible) {
+
+        let apiScanSettingRows = jq('.' + dastApiSetting);
+        jq('.' + dastApiSpecificControls).hide();
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastFreeStyle.js:1051]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 13]
```diff
@@ -0,0 +1,1629 @@
...
+            if (jq('#webSiteNetworkAuthSettingEnabledRow').find('input:checkbox:first').prop('checked') === false) {
+                jq('#webSiteNetworkAuthSettingEnabledRow').find('input:checkbox:first').trigger('click');
+            }
+            let np = jq('#networkPasswordRow').find('input');
+            np.attr('type', 'password');
+        }
+    }
+
+    scanTypeVisibility(isVisible) {
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:528]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 14]
```diff
@@ -0,0 +1,1629 @@
...
+
+                jq('#listWorkflowDrivenAllowedHostUrl').empty();
+                jq('#workflowMacroHosts').val(undefined);
+                jq('#workflowMacroId').val(this.scanSettings.workflowdrivenAssessment.workflowDrivenMacro[0].fileId);
+                this.scanSettings.workflowdrivenAssessment.workflowDrivenMacro[0].allowedHosts.forEach((item, index, arr) => {
+                        console.log(item);
+                        if (arr[index] !== undefined || null) {
+                            jq('#listWorkflowDrivenAllowedHostUrl').append("<li>" + "<input type='checkbox' id=' " + arr[index] +
+                                " ' checked='checked' name='" + arr[index] + "'>" + arr[index] + "</li>");
+                            if (jq('#workflowMacroHosts').val() === null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:634]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 15]
```diff
@@ -0,0 +1,1629 @@
...
+        //only single file upload is allowed from FOD. Todo Iterate the array
+        if (!Object.is(this.scanSettings.workflowdrivenAssessment, undefined)) {
+            if (!Object.is(this.scanSettings.workflowdrivenAssessment.workflowDrivenMacro, undefined)) {
+
+                jq('#listWorkflowDrivenAllowedHostUrl').empty();
+                jq('#workflowMacroHosts').val(undefined);
+                jq('#workflowMacroId').val(this.scanSettings.workflowdrivenAssessment.workflowDrivenMacro[0].fileId);
+                this.scanSettings.workflowdrivenAssessment.workflowDrivenMacro[0].allowedHosts.forEach((item, index, arr) => {
+                        console.log(item);
+                        if (arr[index] !== undefined || null) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:631]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 16]
```diff
@@ -0,0 +1,1629 @@
...
+                            jq('#listWorkflowDrivenAllowedHostUrl').append("<li>" + "<input type='checkbox' id=' " + arr[index] +
+                                " ' checked='checked' name='" + arr[index] + "'>" + arr[index] + "</li>");
+                            if (jq('#workflowMacroHosts').val() === null || undefined) {
+                                jq('#workflowMacroHosts').val(arr[index]);
+                            } else {
+                                let host = jq('#workflowMacroHosts').val();
+                                if (host !== null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:638]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 17]
```diff
@@ -0,0 +1,1629 @@
...
+
+                    }
+                );
+
+            }
+        }
+    }
+
+    scanTypeUserControlVisibility(scanType, isVisible) {
+        if ((isVisible !== undefined || null)) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:656]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 18]
```diff
@@ -0,0 +1,1629 @@
...
+
+    resetLoginMacroSettings() {
+        jq('#loginMacroId').val(undefined);
+        jq('#webSiteLoginMacroEnabled').find('input:checkbox:first').trigger('click');
+    }
+
+    loginMacroSettingsVisibility(isVisible) {
+        let loginMacroSetting = jq('#login-macro-row');
+        let tr = closestRow('#login-macro-row');
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:735]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 19]
```diff
@@ -0,0 +1,1629 @@
...
+            loginMacroSetting.hide();
+            tr.hide();
+        } else {
+            loginMacroSetting.show();
+            tr.show();
+        }
+    }
+
+    excludeUrlVisibility(isVisible) {
+        if ((isVisible === undefined) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:745]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 20]
```diff
@@ -0,0 +1,1629 @@
...
+
+        } else {
+            jq('#standardScanTypeExcludeUrlsRow').show();
+            let tr = closestRow('#standardScanTypeExcludeUrlsRow');
+            tr.show();
+        }
+    }
+
+    timeboxScanVisibility(isVisible) {
+        if ((isVisible === undefined) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:758]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 21]
```diff
@@ -0,0 +1,1629 @@
...
+            jq('#dast-timeBox-scan').hide();
+        } else {
+            jq('#dast-timeBox-scan').show();
+        }
+    }
+
+    networkAuthSettingVisibility(isVisible) {
+        let networkAuth = jq('.dast-networkAuth-setting');
+        if ((isVisible === undefined) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:767]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 22]
```diff
@@ -0,0 +1,1629 @@
...
+            //reset the value here so on changing the scan type the hidden n/w values don't retain the values.
+            this.resetNetworkSettings();
+        } else {
+            networkAuth.show();
+        }
+    }
+
+    websiteScanSettingsVisibility(isVisible) {
+        let standardScanSettingRows = jq('.dast-standard-setting');
+        if ((isVisible === undefined) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:778]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 23]
```diff
@@ -0,0 +1,1629 @@
...
+        jq('[name=webSiteUrl]').val(null);
+        jq('[name=scanTimeBox]').val(null); //timebox not supported for workflow
+
+    }
+
+
+    commonScopeSettingVisibility(isVisible) {
+
+        let commonScopeRows = jq('.scopeContainer');
+        if ((isVisible === undefined) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:807]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 24]
```diff
@@ -0,0 +1,1629 @@
...
+    }
+
+    apiScanSettingVisibility(isVisible) {
+        jq('.dast-api-scan').each((iterator, element) => {
+            let currentElement = jq(element);
+            let tr = closestRow(currentElement);
+            tr.addClass(fodApiScanTypeClassIdr);
+        });
+        let apiScanSettingRows = jq('.dast-api-scan');
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:833]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 25]
```diff
@@ -0,0 +1,1629 @@
...
+
+            apiScanSettingRows.hide();
+        } else {
+            apiScanSettingRows.show();
+        }
+    }
+
+    workflowScanSettingVisibility(isVisible) {
+        let workflowScanSettingRows = jq('.dast-workflow-setting');
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:843]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 26]
```diff
@@ -0,0 +1,1629 @@
...
+            this.resetWorkFlowSettingSValues();
+        } else {
+            workflowScanSettingRows.show();
+            jq('#dast-workflow-macro-upload').show();
+
+        }
+    }
+
+    scanSettingsVisibility(isVisible) {
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:855]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 27]
```diff
@@ -0,0 +1,1629 @@
...
+        }
+    }
+
+    onScanTypeChanged() {
+
+
+        this.resetAuthSettings();
+        let selectedScanTypeValue = jq('#scanTypeList').val();
+
+        if (selectedScanTypeValue === null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:918]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 28]
```diff
@@ -0,0 +1,1629 @@
...
+
+            if (allowedHost !== undefined || null) {
+                if (allowedHost.length > 0) {
+                    allowedHost = allowedHost + "," + hostToAdd;
+                } else
+                    allowedHost = hostToAdd;
+                jq('#workflowMacroHosts').val(allowedHost);
+            }
+        } else {
+            if (allowedHost !== undefined || null) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:1043]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 29]
```diff
@@ -0,0 +1,1629 @@
...
+
+    apiTypeUserControlVisibility(apiType, isVisible) {
+
+        if ((isVisible !== undefined || null)) {
+            this.openApiScanVisibility(false);
+            this.grpcScanVisibility(false);
+            this.graphQlScanVisibility(false);
+            this.postmanScanVisibility(false);
+            switch (apiType) {
+                case "openApi":
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:1306]: Use constant

File: src/main/webapp/DastPipelineGenerator.js [hunk 30]
```diff
@@ -0,0 +1,1629 @@
...
+            jq('#dast-postman').closest('.tr').hide();
+        jq('#dast-openApi').closest('.tr').hide()
+        jq('#dast-graphQL').closest('.tr').hide();
+        jq('#dast-grpc').closest('.tr').hide();
+    }
+
+    apiScanSettingVisibility(isVisible) {
+        let apiScanSettingRows = jq('.dast-api-setting');
+        jq('.dast-api-specific-controls').hide();
+        if ((isVisible === undefined || null) || isVisible === false) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:1378]: Also fix this if statement like I commented above

File: src/main/webapp/DastPipelineGenerator.js [hunk 31]
```diff
@@ -0,0 +1,1629 @@
...
+        );
+    }
+
+    onApiTypeChanged() {
+
+        jq('.uploadMessage').text('');
+        this.onSourceChange('none');
+        let selectedApiTypeValue = jq('#apiTypeList').val();
+
+        if (selectedApiTypeValue === null || undefined) {
```
Comment (c0d3m0nky) [inline src/main/webapp/DastPipelineGenerator.js:1456]: Also fix this if statement like I commented above

**Review by c0d3m0nky — CHANGES_REQUESTED** [review]

**Review by c0d3m0nky — APPROVED** [review]

**Pull request merged**